### PR TITLE
[PW-6545] Error on loading undefined CC data due to invalid flow

### LINKED ISSUE #1440: [PW-6545] Incorrect payment methods
**Describe the bug**
Payment methods do not appear or not correctly.
Paypal is not shown, and apple pay appear ( it should not on Windows/Chrome ).

Retrieve the payment methods response : 
```
{
    "paymentMethodsResponse": {
        "paymentMethods": [
            {
                "brands": [
                    "amex",
                    "maestro",
                    "mc",
                    "visa"
                ],
                "configuration": {
                    "merchantId": "000000000202216",
                    "merchantName": "****"
                },
                "name": "Apple Pay",
                "type": "applepay"
            },
            {
                "brands": [
                    "amex",
                    "maestro",
                    "mc",
                    "visa"
                ],
                "name": "Cartão de crédito",
                "type": "scheme"
            },
            {
                "configuration": {
                    "intent": "capture"
                },
                "name": "PayPal",
                "type": "paypal"
            }
        ]
    },
    "paymentMethodsExtraDetails": {
        "applepay": {
            "icon": {
                "url": "https://checkoutshopper-live.adyen.com/checkoutshopper/images/logos/medium/applepay.png",
                "width": 77,
                "height": 50
            },
            "isOpenInvoice": false,
            "configuration": {
                "amount": {
                    "value": 2845,
                    "currency": "EUR"
                },
                "currency": "EUR"
            }
        },
        "card": {
            "icon": {
                "url": "static.imgix.net/static/version1648806722/base/Magento/base/default/Adyen_Payment/images/logos/card.png",
                "width": 77,
                "height": 50
            },
            "isOpenInvoice": false
        },
        "paypal": {
            "icon": {
                "url": "static.imgix.net/static/version1648806722/base/Magento/base/default/Adyen_Payment/images/logos/paypal.png",
                "width": 65,
                "height": 41
            },
            "isOpenInvoice": false,
            "configuration": {
                "amount": {
                    "value": 2845,
                    "currency": "EUR"
                },
                "currency": "EUR"
            }
        },
        "scheme": {
            "configuration": {
                "amount": {
                    "value": 2845,
                    "currency": "EUR"
                },
                "currency": "EUR"
            }
        }
    }
}
```
Error on the first page of the checkout : 
```
(index):1 Uncaught TypeError: Unable to process binding "foreach: function(){return adyenPaymentMethods }"
Message: Unable to process binding "afterRender: function(){return renderCheckoutComponent() }"
Message: Cannot read properties of undefined (reading 'slice')
    at getStreetAndHouseNumberFromAddress (adyen-hpp-method.min.js:formatted:352:45)
    at UiClass.getFormattedAddress (adyen-hpp-method.min.js:formatted:381:26)
    at UiClass.buildComponentConfiguration (adyen-hpp-method.min.js:formatted:412:49)
    at Object.renderCheckoutComponent (adyen-hpp-method.min.js:formatted:131:46)
    at afterRender (eval at createBindingsStringEvaluator (knockout.min.js:formatted:2593:28), <anonymous>:3:174)
    at evaluateValueAccessor (knockout.min.js:formatted:2786:28)
    at knockout.min.js:formatted:2915:40
    at init (after-render.min.js:1:177)
    at knockout.min.js:formatted:2965:58
    at Object.ignore (knockout.min.js:formatted:1276:45)
getStreetAndHouseNumberFromAddress @ adyen-hpp-method.min.js:formatted:352
getFormattedAddress @ adyen-hpp-method.min.js:formatted:381
buildComponentConfiguration @ adyen-hpp-method.min.js:formatted:412
renderCheckoutComponent @ adyen-hpp-method.min.js:formatted:131
afterRender @ VM7014:3
evaluateValueAccessor @ knockout.min.js:formatted:2786
(anonymous) @ knockout.min.js:formatted:2915
```

The previous error is very weird because of all my tests after that, it do not seems that it enter in this code at the first step, only at the payment step.

It can some from this file `vendor/adyen/module-payment/view/frontend/web/js/model/set-shipping-information-mixin.js`

If we can't find the source of this error, it could help to add a check on the method `getStreetAndHouseNumberFromAddress` maybe ? To prevent this kind of errors.

**To Reproduce**
⚠️ This bug is very hard to reproduce. It seems random.
When this error appear i'm on the first page of the checkout, it's very weird because, from what i saw we should not enter in this code at the first step of the checkout.

The only time a reproduced it, was on a first visit on the website with a clear session.

**Expected behavior**
I should see all the payment methods configured in Magento.

**Magento version**
 2.4.3-p1

**Plugin version**
8.1.1

**Screenshots**
Capture d’écran 2022-04-05 183113
Capture d’écran 2022-04-05 183046

**Desktop (please complete the following information):**
 - OS: Windows / OS X
 - Browser: Chrome
 - Version:  99.0.4844.84

### LINKED ISSUE #1445: [PW-6549] Error in virtuel orders and credit cart in 8.2.0
**Describe the bug**
When we try to checkout with a virtuel order/product it is not possible to select credit card payment due to a js bug

**To Reproduce**
1) Add virtuel product to cart
2) Go to checkout
3) Logo and text is missing for credit card payment and it is not possible to select and checkout
4) JS errors in console

**Magento version**
2.4.3-p1

**Plugin version**
8.2.0

**Screenshots**
Skærmbillede 2022-04-09 kl  18 50 39

### PR DESCRIPTION
**Pull request by tnaber**
**Description**
When the payment methods have not been fully loaded yet, the renderSecureFields method can crash as the checkoutComponent has not been created yet. This PR adds a new observable for the checkoutComponent that will trigger the html to be loaded only if the checkoutComponent is defined. Thus, when the renderSecureFields (now renderCCPaymentMethod) is called the checkoutComponent will always be avaialble

This PR also introduces the adyen-checkout file which abstracts some code shared by the three 'adyen-*-method' files.

**Tested scenario's**
- [x] HPP methods
- [x] Oneclick methods
- [x] CC methods
- [x] 3DS2 modals
- [x] Multishipping
- [x] PayPal

**Fixed issue**:  
Closes #1440
Closes #1445

### CODE REVIEW

File: view/frontend/web/js/view/payment/method-renderer/adyen-hpp-method.js [hunk 1]
```diff
@@ -564,64 +564,68 @@ define(
...
-                } else {
-                    let houseNumber = street.splice(houseNumberStreetLine - 1, 1);
-                    addressArray = {
-                        streetName: street.join(' ').trim(),
-                        houseNumber: houseNumber.join(' ').trim()
+            getFormattedAddress: function (address) {
+                function getStreetAndHouseNumberFromAddress(address, houseNumberStreetLine, customerStreetLinesEnabled) {
+                    if (typeof address.street === 'undefined') {
+                        address.street = [''];
+                    }
```
Comment (tnaber) [inline view/frontend/web/js/view/payment/method-renderer/adyen-hpp-method.js:571]: These are the actually added lines, the rest is reformatting

**Comment by tnaber** [comment]
Hi @jonathanribas,
We're still looking into how to trigger this problem. Does you checkout page allow the payment methods to be loaded without an address being entered yet? E.g. having the payment methods and address detail fields on the same page?

**Comment by jonathanribas** [comment]
Hi @tnaber, we actually use Adobe Commerce checkout with several steps (we don't have a One Step Checkout).

Payment methods are actually triggered by Adyen's module mixin: `vendor/adyen/module-payment/view/frontend/web/js/model/set-shipping-information-mixin.js`, as we have shipping cost estimation on our cart, we get Adyen Payment methods on cart.

If you want you can test it on https://uk.caudalie.com/

**Comment by tnaber** [comment]
Hi @jonathanribas, 
I've found a new solution, it's not pretty so I'm leaving this as a draft for now in order to fully verify. If you are looking for a temporary solution feel free to check it out.

**Comment by jonathanribas** [comment]
Hi @tnaber, thank you! At the moment we will keep our temporary workaround as we don't had other issues since implemented.
Looking forward the final solution from your side, maybe for next release 8.2.2?

**Comment by jonathanribas** [comment]
Hi @tnaber, with those changes, do you still need to use `afterRender` on `cc-form.html`?

**Comment by tnaber** [comment]
We might, as the field does not necessarily have to be available when the loadCheckoutComponent is ran. E.g. if we delete the afterRender and the loadCheckoutComponent function is ran for the last time before the field is available it will never retrigger the mount. 
However, I do agree that a single trigger would be neater and I'm still testing if we can do without one of both

**Comment by kbakkes** [comment]
Hi, We are currently facing the same as described above.

**Comment by tnaber** [comment]
Hi @kbakkes and @jonathanribas,

I have updated the fix and implemented a new flow. I am confident this should fix the issue.
If you would be able to test it that would be greatly appreciated!

Cheers,
Titus

File: view/frontend/web/js/view/payment/method-renderer/adyen-cc-method.js [hunk 2]
```diff
@@ -70,50 +70,55 @@ define(
...
+                    paymentMethodsResponse,
+                    this.handleOnAdditionalDetails.bind(this)
+                )
+
+                if (!!this.checkoutComponent) {
+                    // Setting the adyenCCMethod will load the html and trigger the renderPaymentMethod on afterRender
+                    self.adyenCCMethod({
+                            icon: !!paymentMethodsResponse.paymentMethodsExtraDetails.card.icon
+                                ? paymentMethodsResponse.paymentMethodsExtraDetails.card.icon
+                                : {}
```
Comment (michaelpaul) [inline view/frontend/web/js/view/payment/method-renderer/adyen-cc-method.js:118]:
```suggestion
                            icon: paymentMethodsResponse.paymentMethodsExtraDetails.card.icon || {}
```

**Review by michaelpaul — CHANGES_REQUESTED** [review]

**Review by michaelpaul — APPROVED** [review]

**Review by peterojo — APPROVED** [review]

**Pull request merged**